src/core/coap/coap.cpp

```diff
@@ -144,17 +144,20 @@ otError CoapBase::SendMessage(Message &               aMessage,
     Message *    storedCopy = NULL;
     uint16_t     copyLength = 0;
 
-    if ((aMessage.GetType() == OT_COAP_TYPE_ACKNOWLEDGMENT || aMessage.GetType() == OT_COAP_TYPE_RESET) &&
-        aMessage.GetCode() != OT_COAP_CODE_EMPTY)
+    switch (aMessage.GetType())
     {
+    case OT_COAP_TYPE_ACKNOWLEDGMENT:
+        assert(aMessage.IsMessageIdValid());
         mResponsesQueue.EnqueueResponse(aMessage, aMessageInfo);
-    }
-
-    // Set Message Id if it was not already set.
-    if (aMessage.GetMessageId() == 0 &&
-        (aMessage.GetType() == OT_COAP_TYPE_CONFIRMABLE || aMessage.GetType() == OT_COAP_TYPE_NON_CONFIRMABLE))
-    {
+        break;
+    case OT_COAP_TYPE_RESET:
+        assert(aMessage.IsMessageIdValid());
+        assert(aMessage.GetCode() == OT_COAP_CODE_EMPTY);
+        break;
+    default:
+        assert(!aMessage.IsMessageIdValid());
```

Comment:
I think this assert might fail because CON/NON COAP messages  can also have message ID already set. 
For example, `case OT_COAP_TYPE_NON_CONFIRMABLE` in `CoapBase::SendHeaderResponse`

Reply:
I found that in the end, seems we don't actual need to check `IsMessageIdValid` at all...